source/extensions/health_checkers/mysql/mysql.cc

```diff
@@ -0,0 +1,323 @@
+  constexpr const size_t ok_response_size = 1;
+  constexpr const size_t ok_packet_size_threshold = 7;
+
+  const size_t minimum_needed_packet_size = mysql_packet_header_size + ok_response_size;
+  if (data.length() < minimum_needed_packet_size) {
+    ENVOY_CONN_LOG(trace,
+                   "mysql healthcheck failed [response]: packet truncated, expected at least {} "
+                   "bytes, got {} ",
+                   *client_, minimum_needed_packet_size, data.length());
+    return false;
+  }
+
+  const auto packet = reinterpret_cast<const uint8_t*>(data.linearize(minimum_needed_packet_size));
+  const auto payload_length =
+      read_little_endian_int<mysql_payload_length_size>(packet + mysql_payload_length_offset);
+  if (data.length() != mysql_packet_header_size + payload_length) {
+    ENVOY_CONN_LOG(
+        trace, "mysql healthcheck failed [response]: payload length mismatch, expected {}, got {} ",
+        *client_, mysql_packet_header_size + payload_length, data.length());
+    return false;
+  }
+
+  const auto sequence_id =
+      read_little_endian_int<mysql_sequence_id_size>(packet + mysql_sequence_id_offset);
+  if (sequence_id != 2) {
+    ENVOY_CONN_LOG(trace, "mysql healthcheck failed [response]: invalid sequence id {} ", *client_,
+                   sequence_id);
+    return false;
+  }
+
+  const auto response_code =
+      read_little_endian_int<ok_response_size>(packet + mysql_payload_offset);
+  if (response_code != MYSQL_OK_RESPONSE || data.length() < ok_packet_size_threshold) {
+    ENVOY_CONN_LOG(trace,
+                   "mysql healthcheck failed [response]: expected an OK ({}) response with minimum "
+                   "packet size {{}}, got {} with size {} ",
+                   *client_, MYSQL_OK_RESPONSE, ok_packet_size_threshold, response_code,
+                   data.length());
+    return false;
+  }
+
+  return true;
+}
+
+void MySQLHealthChecker::MySQLActiveHealthCheckSession::onData(Buffer::Instance& data) {
+  switch (phase_) {
+  case Phase::Greeting: {
+    if (!parse_mysql_server_greeting_packet(data)) {
+      break;
+    }
+
+    data.drain(data.length());
+    phase_ = Phase::Reply;
+
+    Buffer::OwnedImpl buffer;
+    write_mysql_login_request_packet(buffer);
+    write_mysql_quit_packet(buffer);
+    client_->write(buffer, false);
+    return;
+  }
+
+  case Phase::Reply: {
+    if (!parse_mysql_ok_packet(data)) {
+      break;
+    }
+
+    phase_ = Phase::Over;
+    data.drain(data.length());
+
+    handleSuccess();
+    if (!parent_.reuse_connection_) {
+      client_->close(Network::ConnectionCloseType::NoFlush);
+    }
+    return;
+  }
+
+  case Phase::Over:
+    ENVOY_CONN_LOG(trace, "unexpected data received from mysql host", *client_);
+    data.drain(data.length());
+    return;
+  }
+
+  phase_ = Phase::Over;
+  data.drain(data.length());
+
+  handleFailure(FailureType::Active);
+  client_->close(Network::ConnectionCloseType::NoFlush);
+}
```

Comment:
A whole lot of functions here can be useful for anyone wanting to build a full blown MySQL codec in Envoy. I suggest moving these useful MySQL parsing stuff (login, packet okay and others) into a generic MySQL folder (I don’t know the right destination for it given that there is no MySQL filter), but it would provide the filter implementor a jump start. With all the codec functions abstracted, you could encapsulate checks for partial packets etc into the codec, such that the health check is dramatically simplified into 
 MySQL.doLogin(user)
 MySQL.resp == okay -> health check ok

Reply:
sounds like a good idea. I'm not sure that would be useful out-of-the-box for a filter implementor, but it's a starting skeleton as you mentioned. I'll see what I can come up with